compiler/rustc_trait_selection/src/transmute.rs

```diff
@@ -0,0 +1,78 @@
+use rustc_infer::infer::InferCtxt;
+use rustc_infer::traits::{Obligation, ObligationCause};
+use rustc_middle::ty::{Binder, ParamEnv, PolyTraitRef, Ty, TyCtxt};
+use rustc_transmute::layout::rustc::Ref;
+use rustc_transmute::{Answer, Assume, Condition, Reason};
+
+/// Attempts to confirm the given transmutability candidate.
+///
+/// The return type of this routine reflects that it is designed to support the
+/// needs of both trait selection and error reporting; it returns:
+/// - `Ok(None)` if the types are transmutable
+/// - `Ok(Some(src, dst, assumptions, condition))` if the `src` is transmutable
+///   into `dst` when `condition` holds under `assumptions`
+/// - `Err(None)` if transmutability cannot be assessed; e.g., due to a
+///   malformed `Assume`
+/// - `Err(Some((src, dst, reason))` if `src` is not transmutable into `dst`
+///   because of `reason`.
+pub(crate) fn confirm_transmutability_candidate<'tcx, P>(
+    infcx: &InferCtxt<'tcx>,
+    obligation: &Obligation<'tcx, P>,
+    trait_ref: PolyTraitRef<'tcx>,
+) -> Result<
+    Option<(Ty<'tcx>, Ty<'tcx>, Assume, Condition<Ref<'tcx>>)>,
+    Option<(Ty<'tcx>, Ty<'tcx>, Reason<Ref<'tcx>>)>,
+> {
+    /// Attempts to deeply normalize `ty`.
+    fn try_normalize<'tcx>(
+        tcx: TyCtxt<'tcx>,
+        param_env: ParamEnv<'tcx>,
+        ty: Binder<'tcx, Ty<'tcx>>,
+    ) -> Option<Binder<'tcx, Ty<'tcx>>> {
+        use rustc_infer::infer::TyCtxtInferExt;
+
+        use crate::traits::ObligationCtxt;
+        let infcx = tcx.infer_ctxt().with_next_trait_solver(true).build();
+        let ocx = ObligationCtxt::new(&infcx);
+        let cause = ObligationCause::dummy();
+        let Ok(ty) = ocx.deeply_normalize(&cause, param_env, ty) else { return None };
+        let errors = ocx.select_all_or_error();
+        if errors.is_empty() { Some(ty) } else { None }
+    }
+
+    let dst = trait_ref.map_bound(|tr| tr.args.type_at(0));
+    let src = trait_ref.map_bound(|tr| tr.args.type_at(1));
```

Comment:
It seems a bit odd to first map the binder, then skip it immediately after. You should instantiate this binder with placeholders.